[DeviceMesh] Rename _device_mesh.py to device_mesh.py to prepare for beta (#115099)

wz337 opened the pull request:
Summary:

Rename _device_mesh.py to device_mesh.py, update all callsites, add documentation.
We created stubs for public class and methods in torch.distributed.device_mesh so that torch.distributed.device_mesh can be imported with or without distributed is available().


Original diff reverted: D51629761
Original PR reverted: https://github.com/pytorch/pytorch/pull/115099
Prior to landing, CI signals are all passed. Shipit added the "ci/trunk" label to the PR and DID NOT wait for it and went ahead committing. More context can be found in the reverted PR above. 

Test Plan: CI.

Differential Revision: D51861018



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @kiukchung @lucasllc

malfet commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -1,4 +1,4 @@
-from torch.distributed._device_mesh import (  # noqa: F401
+from torch.distributed.device_mesh import (  # noqa: F401
```
We should probably return early if `torch.distribtued.is_avaialble()` is False, to satisfy public_api test

wz337 committed:
[DeviceMesh] Rename _device_mesh.py to device_mesh.py to prepare for beta (#115193)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/115193

Rename _device_mesh.py to device_mesh.py, update all callsites, add documentation.
We created stubs for public class and methods in torch.distributed.device_mesh so that torch.distributed.device_mesh can be imported with or without distributed is available().

Original diff reverted: D51629761
Original PR reverted: https://github.com/pytorch/pytorch/pull/115099
Prior to landing, CI signals are all passed. Shipit added the "ci/trunk" label to the PR and DID NOT wait for it and went ahead committing. More context can be found in the reverted PR above.

Test Plan: CI.

Differential Revision: D51861018

fbshipit-source-id: dc7b26cea7340d55498730123e82a42cef46ff55

fegin reviewed (approved):
The version seems to pass all the tests now.

wz337 commented:
@pytorchmergebot merge
